Remove requirement for patched TFT_eSPI library

bradmck opened the pull request:
This pull request removes the requirement to use a patched (and older) TFT_eSPI library.  The external_component will just pull the latest TFT_eSPI library.  The TFT_eSPI settings required to make the display on the LilyGo T-Display-S3 work are set within the `display.py` file.
NOTE: I've updated the example.yaml and example-with-touch.yaml files to point to my repository for testing.  We'll want to point those back to your repo @landonr if this PR is approved and merged.

patfelst commented:
Hi @bradmck don't want to nag you, but how are you going with exposing the TFT_eSPI object? Is there a branch somewhere that I can use this? thanks!

bradmck commented:
Hey @patfelst - it's no bother.  I've run into some issues around trying to get rotation working correctly.  I had it working as long as you didn't specify rotation on the display component and instead used setRotation in the on_boot section, but that was hacky and confusing.  So I'm trying to figure out how to get the rotations between ESPHome and TFT_eSPI to match while not messing up the sprite.  I'm still working on it and hoping to make progress in the next few days.

guillempages commented:
Hi @bradmck I've just tried again replacing landonr's repo with yours on my working setup, and I cannot see anything drawn on the screen :-(
Touch is working, the device is responsive, and I can still turn on and off the backlight, but any standard ESPHome display drawing commands seem to be ignored (I'm using both `printf`, `image` (both static images and animations)  and `rectangle`).

Could it be, that there is some difference on the display between the touch version and the no-touch version?

guillempages commented on components/tdisplays3/display.py:
```diff
@@ -28,10 +28,45 @@ def validate_tdisplays3(config):
+    # Add platformio build_flags for the correct TFT_eSPI settings for the T-Display-S3
+    # This allows using current, unpatched versions of TFT_eSPI
+    cg.add_build_flag("-DUSER_SETUP_LOADED=1")
+    cg.add_build_flag("-DST7789_DRIVER=1")
```
The TFT_eSPI define does not have a value
```suggestion
    cg.add_build_flag("-DST7789_DRIVER")
```

guillempages commented on components/tdisplays3/display.py:
```diff
@@ -28,10 +28,45 @@ def validate_tdisplays3(config):
+    # This allows using current, unpatched versions of TFT_eSPI
+    cg.add_build_flag("-DUSER_SETUP_LOADED=1")
+    cg.add_build_flag("-DST7789_DRIVER=1")
+    cg.add_build_flag("-DINIT_SEQUENCE_3=1")
```
The TFT_eSPI define does not have a value
```suggestion
    cg.add_build_flag("-DINIT_SEQUENCE_3")
```

guillempages commented on components/tdisplays3/display.py:
```diff
@@ -28,10 +28,45 @@ def validate_tdisplays3(config):
+    cg.add_build_flag("-DINIT_SEQUENCE_3=1")
+    cg.add_build_flag("-DCGRAM_OFFSET")
+    cg.add_build_flag("-DTFT_RGB_ORDER=TFT_RGB")
+    cg.add_build_flag("-DTFT_INVERSION_ON=1")
```
The TFT_eSPI define does not have a value
```suggestion
    cg.add_build_flag("-DTFT_INVERSION_ON")
```

guillempages commented on components/tdisplays3/display.py:
```diff
@@ -28,10 +28,45 @@ def validate_tdisplays3(config):
+    cg.add_build_flag("-DCGRAM_OFFSET")
+    cg.add_build_flag("-DTFT_RGB_ORDER=TFT_RGB")
+    cg.add_build_flag("-DTFT_INVERSION_ON=1")
+    cg.add_build_flag("-DTFT_PARALLEL_8_BIT=1")
```
The TFT_eSPI define does not have a value
```suggestion
    cg.add_build_flag("-DTFT_PARALLEL_8_BIT")
```

guillempages commented on components/tdisplays3/display.py:
```diff
@@ -28,10 +28,45 @@ def validate_tdisplays3(config):
+    cg.add_build_flag("-DTFT_D6=47")
+    cg.add_build_flag("-DTFT_D7=48")
+    cg.add_build_flag("-DLOAD_GLCD=1")
+    cg.add_build_flag("-DLOAD_FONT2=1")
+    cg.add_build_flag("-DLOAD_FONT4=1")
+    cg.add_build_flag("-DLOAD_FONT6=1")
+    cg.add_build_flag("-DLOAD_FONT7=1")
+    cg.add_build_flag("-DLOAD_FONT8=1")
+    cg.add_build_flag("-DLOAD_GFXFF=1")
+    cg.add_build_flag("-DSMOOTH_FONT=1")
```
Those defines seem related to fonts; which shouldn't be relevant when using ESPHome fonts. Could you test with your setup, whether it works without those defines?
```suggestion

```

guillempages commented on components/tdisplays3/display.py:
```diff
@@ -28,10 +28,45 @@ def validate_tdisplays3(config):
+    cg.add_build_flag("-DTFT_D5=46")
+    cg.add_build_flag("-DTFT_D6=47")
+    cg.add_build_flag("-DTFT_D7=48")
+    cg.add_build_flag("-DLOAD_GLCD=1")
```
This is probably also related to fonts (see next comment); you might want to try whether it is really needed.
```suggestion
    cg.add_build_flag("-DLOAD_GLCD=1")
```

bradmck replied:
Hi @guillempages - I'm still puzzled that yours doesn't display anything.  Thanks for the suggestions on define cleanup.  On the font defines - I may be using them in my next version, but can look into pulling them out of this version.  I'll continue to play around with this.  I guess it's possible there's something different between the touch and non-touch versions - however the first version I came up with worked for you.  There's not much different here except this version doesn't expose the TFT and SPR objects and the defines are now handled in the Python file.  

I might come up with a special branch for you to try that would just enable some debug logging around the display driver that is being loaded.  But let's definitely hold on merging this PR since it's still not working for @guillempages.

guillempages reviewed (commented):
The suggestions I made are generic suggestions; they still don't fix the issue that my display doesn't show anything. I've even tried the example.yaml files but no luck. Touch and backlight work; the display doesn't draw anything :-(

patfelst commented:
> Hey @patfelst - it's no bother. I've run into some issues around trying to get rotation working correctly. I had it working as long as you didn't specify rotation on the display component and instead used setRotation in the on_boot section, but that was hacky and confusing. So I'm trying to figure out how to get the rotations between ESPHome and TFT_eSPI to match while not messing up the sprite. I'm still working on it and hoping to make progress in the next few days.

No worries @bradmck I wasn't sure if you were still working on it. To be honest I'd be perfectly happy to set the rotation in the display lambda, assuming that works. This probably deviates from a "proper" ESPHome component rule, but at least we could start developing with sprites etc. Thanks for your great work!

guillempages commented:
@bradmck I'm not 100% sure that I got any of your versions working; I never tried the ones with the defines in the yaml file.

@sbur83 also has the touch display version; would you be so kind and try bradmck's version, to see if it is a problem with my setup or a more generic problem with the touch version? In theory, just changing the repo name on the external_components for your display should do the trick.

landonr commented:
this worked for me without touch

sbur83 commented:
Sorry guys, I'll get to this tomorrow arvo I keep forgetting. Landon's setup is working fine with rotation 170. I haven't updated it for a while. I'm off to bed and will test bradmck's tomorrow arvo

sbur83 commented:
> 

I had the black screen with backlight only/ grey screen. Restarting esp device would not fix it and I was also convinced that the bradmck branch was broken. By chance I had to restart homeassistant and after restarting the display was working again and still pointing to bradmck repo. I guess it might be something getting cached in esphome folder, don't know but it is working after restart. Maybe @guillempages can reproduce the error still and try restart esphome addon?

clowrey commented:
Was that with the demo that only displays the time? Maybe just something with the time not being updated from home assistant - we should add a few other elements to the default YAML example like some boxes etc I think.

sbur83 commented:
@clowrey not the demo, I am displaying date and time from homeassistant. Also working with other homeassistant sensors and text sensors

guillempages commented:
@bradmck I have some good news!
After reading @sbur83's experience by restarting homeassistant I tried cleaning up the ".esphome/build" directory (I'm not using the homeassistant add-on; I'm directly using the esphome CLI). That didn't help either, and I thought "let's try to flash the image via USB; maybe that's the problem". To do that I had to unplug the USB connector and plug it again, and surprise: after re-plugging it, it showed the expected content.

So the upstream TFT_eSPI is also working for me now :-) It was a case of "Have you tried plugging it off and on again" :-D

The only thing I noticed is that the colors seem different; I was using (255, 200, 0) for an orange color, and now it is much more yellowish. But that's something I can live with.

If I could wish for something on your implementation though, would be to *not* enable the font defines by default, so that people using plain esphome [i.e. without directly accessing the tft methods] don't need to pay the extra flash cost of adding the TFT_eSPI font support.

bradmck commented:
Hey that's great @guillempages !  I was really confused why this wasn't working for you.  I'm a bit concerned that the color is off for you - I don't know why that would be off the top of my head.  My code does set the color order to TFT_RGB which works fine for me, but maybe the touch is different?  Thank you very much for your patience and for testing multiple times.

On having to unplug -- I got bit by this myself.  I had code I was testing that was working great and I thought I was good.  Until I unplugged/re-plugged.  It turns out that running the code from @landonr's repo first had properly initialized the display so my code worked until power cycle.  I did get that sorted out though.

I don't have any problem re-working the defines to disable the font stuff - as you requested in the feedback above.  They probably aren't needed unless working with the TFT/SPR objects so it makes sense to do it your way.  I'll do a bit of testing on that and modify accordingly.  There's also the small detail that the example.yaml is pointing to my repo - what's the best way to merge this but get those changed back to @landonr's repo?

bradmck commented:
@guillempages - I've committed your suggestions except for the one for removing `cg.add_build_flag("-DLOAD_GLCD=1")`.  I couldn't commit that from here since it doesn't actually make a code change, so I just updated my repo manually with that suggestion.  All font defines are now removed and it's working fine for me.  Thank you for the suggestions.

guillempages commented on example-with-touch.yaml:
```diff
@@ -2,7 +2,7 @@ esphome:
 
 external_components:
-  - source: github://landonr/lilygo-tdisplays3-esphome
+  - source: github://bradmck/lilygo-tdisplays3-esphome
```
```suggestion
  - source: github://landonr/lilygo-tdisplays3-esphome
```

guillempages commented on example.yaml:
```diff
@@ -2,7 +2,7 @@ esphome:
 
 external_components:
-  - source: github://landonr/lilygo-tdisplays3-esphome
+  - source: github://bradmck/lilygo-tdisplays3-esphome
```
```suggestion
  - source: github://landonr/lilygo-tdisplays3-esphome
```

guillempages reviewed (commented):
Looks good to me.

The two suggestions are to make sure that the "official" (i.e. @landonr's) repository is mentioned in the examples.

patfelst commented:
Hi @guillempages, regarding your comment:
> The only thing I noticed is that the colors seem different; I was using (255, 200, 0) for an orange color, and now it is much more yellowish. But that's something I can live with.

I just stumbled across this https://github.com/Bodmer/TFT_eSPI/discussions/2455, somebody else has noticed the colors are off on this display. Seems no solution yet, but there is a linked issue on the [Lilygo page](https://github.com/Xinyuan-LilyGO/T-Display-S3/issues/113) that would be worth watching.

guillempages commented:
@patfelst I've just tried putting a red rectangle and yes, it is very orange'ish. As I said, this is something I can live with (and I also think that the issue is in the TFT_eSPI library, or even in the display driver itself). That might be the reason why the example Lilygo repository is using a patched version of the TFT_eSPI library, instead of the plain upstream one.
